### DISCUSSION TITLE
Invoke-pester When Powershell file test is in a Function seems to not locate any tests

### DISCUSSION THREAD

**Discussion by rcoughtrey** [pester/Pester#2156] [Q&A] [answered]
I have been reading and reading but just cannot pin point an answer to my issue or lack of knowledge.

Issue Outline:
When calling my tests with pester everything runs but no tests are ever found. I have dug around to find out that If the *.test.ps1 file being called is inside a powershell Function nothing is found. The moment I remove the function everything works as expected.

Here is a cut back version of my test sample:

Calling the invoke-pester

```powershell
$Container = New-PesterContainer -Path "C:\QE\Pester\PesterStart.Test.ps1"

$Configuration = New-PesterConfiguration
$Configuration.Run.Container = $Container
$Configuration.TestResult.Enabled= $true
$Configuration.TestResult.OutputFormat = "NUnitXML"
$Configuration.TestResult.OutputPath = "Test-Pester.XML"
$Configuration.Output.Verbosity = "Detailed"

Invoke-Pester -Configuration $Configuration
```

===========
PesterStart.Test.ps1 file:

```powershell
function Set-location
{
    [CmdletBinding()]
    Param
    (
        [Parameter(Mandatory=$False)][String]$NewCodes        
    )

BeforeAll{
Import-Module "C:\QE\Pester\Location.ps1" -Force
}

Describe 'Get urlLocation' {
    Context 'urlLocation Creation' {
        It "Auction Was set to east" {
            Set-location -Location "east" | Should -Be "east"
        }
    }
}
}
```

This will run but I get the following:
`Starting discovery in 1 files.
Discovery found 0 tests in 62ms.
Running tests.
Tests completed in 60ms
Tests Passed: 0, Failed: 0, Skipped: 0 NotRun: 0`

Now if I change PesterStart.Test.ps1 file to:

```powershell
BeforeAll {
Import-Module "C:\Users\266542\OneDrive - CarMax\QE\Pester\Location.ps1" -Force
}

Describe 'Get urlLocation' {
    Context 'urlLocation Creation' {
        It "Auction Was set to east" {
            Set-location -Location "east" | Should -Be "east"
        }
    }
}
```

My output is

```
Starting discovery in 1 files.
Discovery found 1 tests in 131ms.
Running tests.

Running tests from 'C:\Users\266542\OneDrive - CarMax\QE\Pester\PesterStart.Test.ps1'
Describing Get urlLocation
 Context urlLocation Creation
   [+] Auction Was set to east 5ms (2ms|3ms)
Tests completed in 337ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 0
```

Any ideas :)

**Comment by rcoughtrey** [comment]
I have worked around my issue, I just remove the function from the first script that contained the tests. Just had make some changes in azure to support how that script/function was being called.

It still would be good to know why invoke-pester cannot call a script which has its tests in a function

**Reply by fflaten** [reply]
You never called the function inside the pesterstart.tests.ps1 file. 

Either way, having trouble understanding why you'd want this at all.

**Comment by rcoughtrey** [comment]
Calling the functions pesterstart.tests.ps1 to start the tests does not have issue. Its just that the script that is passed into invoke-pester cannot have a powershell function. I removed the function and just used params to get the data into the test script

**Comment by nohwnd** [comment, marked as answer]
Pester runs the file to find the tests, your file only defines a function, it never calls it, so `Describe` never runs and discovery finds nothing.

A test file can have functions in it, you just need the `Describe` blocks to run when the file is invoked. Either move them out of the function, or call the function at the end of the file.
